### PR TITLE
WIP#19 Disable/enable components based on status

### PR DESCRIPTION
**Pull request by ellienguyen**
- Add `REFRESHED` status 
- Add `setEnabled` method to `TitledPanel`
- Separate every panel into `MainView` scope
- Add `setEnabledChildren` to `MainView` to easily disable/enable every panel

Not done:
- Waiting for `SheetFreshController` to be pulled in to add set `REFRESHED` status

### CODE REVIEW

File: src/main/java/org/vietabroader/view/MainView.java [hunk 1]
```diff
@@ -47,6 +47,15 @@ private TitledPanel(String title) {
                     new EmptyBorder(8, 2, 8, 2))    // Inner padding of each panel
             );
         }
+
+        @Override
+        public void setEnabled(boolean b) {
+            this.setEnabled(b);
```
Comment (ctring) [inline src/main/java/org/vietabroader/view/MainView.java]: use `super.setEnabled()`. Calling `this.setEnabled()` means calling this method, not the parent's one, making an indefinite recursive loop.

File: src/main/java/org/vietabroader/view/MainView.java [hunk 2]
```diff
@@ -408,25 +428,43 @@ public void update(Observable o, Object arg) {
...
+                setEnabledChildren(false);
+                break;
+            case REFRESHED:
+                setEnabledChildren(true);
+                panWebcam.setEnabled(true);
                 break;
         }
     }
+
+    public void setEnabledChildren(boolean b) {
```
Comment (ctring) [inline src/main/java/org/vietabroader/view/MainView.java]: Good!

File: src/main/java/org/vietabroader/view/MainView.java [hunk 3]
```diff
@@ -408,25 +428,43 @@ public void update(Observable o, Object arg) {
...
-
+                setEnabledChildren(true);
                 List<String> sheets = currentSpreadsheet.getSheetTitles();
                 cbbSheet.removeAllItems();
                 sheets.forEach(cbbSheet::addItem);
                 break;
             case QR_READING:
+                setEnabledChildren(false);
+                break;
+            case REFRESHED:
```
Comment (ctring) [inline src/main/java/org/vietabroader/view/MainView.java]: Move this above `case QR_READING`

File: src/main/java/org/vietabroader/view/MainView.java [hunk 4]
```diff
@@ -408,25 +428,43 @@ public void update(Observable o, Object arg) {
...
+                panWebcam.setEnabled(true);
                 break;
         }
     }
+
+    public void setEnabledChildren(boolean b) {
+        panSignIn.setEnabled(b);
+        panSpreadsheet.setEnabled(b);
+        panWorkspace.setEnabled(b);
+        panWebcam.setEnabled(b);
```
Comment (ctring) [inline src/main/java/org/vietabroader/view/MainView.java]: Let's comment this line out for now. We're still working on the camera controller so it's pretty inconvenient to connect to a sheet every time we need to test the webcam.

**Review by ctring — APPROVED** [review]
At the first line of `createColumnPanel()`, modify it to be like this:

```
JPanel panelMain = new JPanel() {
            @Override
            public void setEnabled(boolean b) {
                super.setEnabled(b);
                Component[] children = this.getComponents();
                for (Component child: children) {
                    child.setEnabled(b);
                }
            }
        };
```
It's patchy but it works so...

You are good to go after these changes

File: src/main/java/org/vietabroader/view/MainView.java (hunk 2 above)
Comment (ctring) [reply src/main/java/org/vietabroader/view/MainView.java]: Use `private` since we don't have the intention to use this method outside of this class.

**Review by ctring — APPROVED** [review]
Add this line after `GlobalState.getInstance().addObserver(view);` in `Main.java`:
```            
GlobalState.getInstance().setStatus(GlobalState.Status.SIGNED_OUT);
```

**Comment by ellienguyen** [comment]
List of things changed
- Add `SIGNED_OUT` after closing windows
- Add conditions for `REFRESHED` state in `SheetRefreshController` and `SpreadsheetConnectController`
- Set the background of refresh label to normal after successfully refresh

**Pull request merged**